packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -0,0 +1,519 @@
+  | {
+      type: 'Collection';
+      namespace: string;
+      collectionType: 'collection' | 'view' | 'timeseries' | null;
+      activeSubTab:
+        | 'Documents'
+        | 'Aggregations'
+        | 'Schema'
+        | 'Indexes'
+        | 'Validation'
+        | null;
+      initialQuery?: unknown;
+      initialAggregation?: unknown;
+      initialPipelineText?: string;
+      editingViewNamespace?: string;
+
+      // TODO: make collection-tab plugin resolve metadata on its own
+      metadata: CollectionTabPluginMetadata;
+    };
+
+export type WorkspaceByType<T extends Workspace['type']> = Extract<
+  Workspace,
+  { type: T }
+>;
+
+export type WorkspaceTab = { id: string } & Workspace;
+
+export type WorkspacesState = {
+  tabs: WorkspaceTab[];
+  activeTabId: string | null;
+};
+
+const getTabId = () => {
+  return new ObjectId().toString();
+};
+
+export const getInitialTabState = (
+  workspace: OpenWorkspaceOptions
+): WorkspaceTab => {
+  const tabId = getTabId();
+  switch (workspace.type) {
+    case 'Databases':
+    case 'My Queries':
+    case 'Performance':
+      return {
+        id: tabId,
+        type: workspace.type,
+      };
+    case 'Collections':
+      return {
+        id: tabId,
+        type: workspace.type,
+        namespace: workspace.namespace,
+      };
+    case 'Collection': {
+      const activeSubTab =
+        workspace.initialAggregation || workspace.initialPipelineText
+          ? 'Aggregations'
+          : workspace.initialQuery
+          ? 'Documents'
+          : workspace.activeSubTab ?? 'Documents';
+
+      // TODO: make collection-tab plugin resolve metadata on its own
+      const collectionType = workspace.metadata.isTimeSeries
+        ? 'timeseries'
+        : workspace.metadata.isReadonly
+        ? 'view'
+        : 'collection';
+
+      return {
+        id: tabId,
+        type: workspace.type,
+        namespace: workspace.namespace,
+        collectionType,
+        activeSubTab,
+        initialQuery: workspace.initialQuery,
+        initialAggregation: workspace.initialAggregation,
+        initialPipelineText: workspace.initialPipelineText,
+        editingViewNamespace: workspace.editingViewNamespace,
+
+        // TODO: make collection-tab plugin resolve metadata on its own
+        metadata: workspace.metadata,
+      };
+    }
+  }
+};
+
+const getInitialState = () => {
+  return {
+    tabs: [] as WorkspaceTab[],
+    activeTabId: null,
+  };
+};
+
+const reducer: Reducer<WorkspacesState> = (
+  state = getInitialState(),
+  action
+) => {
+  if (isAction<OpenWorkspaceAction>(action, WorkspacesActions.OpenWorkspace)) {
+    if (action.newTab) {
+      const newTab = getInitialTabState(action.workspace);
+      return {
+        tabs: [...state.tabs, newTab],
+        activeTabId: newTab.id,
+      };
+    }
+
+    if (
+      action.workspace.type === 'My Queries' ||
+      action.workspace.type === 'Databases' ||
+      action.workspace.type === 'Performance'
```

Comment:
I'll be moving tests from tabs.ts store to this one, in the meantime, the gist for the tab behavior is that we're trying to preserve the current tab / workspace relation:

If you open a new tab for the instance stuff, we'll always try to find an open one and switch to it, this matches the current compass behavior where we don't really have any way to open any of those in the new tab.

If you open anything that requires namespace: database collections list or collection tab, we'll replace the current active tab content with the new namespace tab, but only if the current tab type matches the new one that we're trying to open.

For example: if you'd try to open a collection while looking at "my queries" tab, this will open a new tab with the selected collection namespace. But if it's a collection tab for a different namespace that is active, we'll replace the content, same way it works right now.

For this project we're just keeping the behavior as close to the current one as possible with the change. Some changes might be made in the follow up project that is more focused on the UI part of compass web.